src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -660,6 +663,32 @@ ExtentPlacementManager::BackgroundProcess::do_background_cycle()
   }
 }
 
+// adjust generations located in [MIN_REWRITE_GENERATION, MIN_COLD_GENERATION)
+void ExtentPlacementManager::BackgroundProcess::update_generation_mappings()
+{
+  if (!cold_cleaner) {
+    return;
+  }
+
+  auto stat = major_cleaner->get_stat();
+  double used_ratio = (double)stat.data_stored / (double)stat.total;
+  if (used_ratio > start_evict_ratio && !start_evict) {
+    start_evict = true;
+    for (rewrite_gen_t gen = MIN_REWRITE_GENERATION;
+         gen < MIN_COLD_GENERATION;
+         ++gen) {
+      generation_mappings[gen] = MIN_COLD_GENERATION;
+    }
+  } else if (used_ratio < stop_evict_ratio && start_evict) {
+    start_evict = false;
+    for (rewrite_gen_t gen = MIN_REWRITE_GENERATION;
+         gen < MIN_COLD_GENERATION;
+         ++gen) {
+      generation_mappings[gen] = gen;
```

Comment:
Seems there are two modes to control the eviction above.

1. start_evict == true: evict all `target-gen` [MIN_REWRITE_GENERATION, MIN_COLD_GENERATION) as `gen` MIN_COLD_GENERATION
2. start_evict = false: evict the `target-gen` MIN_COLD_GENERATION as `gen` MIN_COLD_GENERATION

Looks to me that `2. start_evict = false` will still evict extents when the cold tier is present.

According to the discussion https://github.com/ceph/ceph/pull/48762#discussion_r1016296985, it is intuitive to implement a 3rd mode without eviction, quoted: 
```
When the hot tier is empty, EPM can assign the target gen(5) extents back to gen(4) to retain them in the hot tier, which is identical to the existing behavior without a cold tier.
```
That is:
3. no-eviction: set the `target-gen` MIN_COLD_GENERATION back to `gen` MIN_COLD_GENERATION - 1

That said, we can introduce 3 modes to control the evictions, roughly may look like:
```c++
// in class BackgroundProcess

enum class eviction_mode_t {
  STOP,     // set to mode 3 above
  DEFAULT,  // set to mode 2 above
  FAST      // set to mode 1 above
};

eviction_mode_t eviction_mode = eviction_mode_t::STOP;

// It seems that get_generation_mapping() is less explicit than the direct logic here:
rewrite_gen_t adjust_generation_with_eviction(rewrite_gen_t gen) {
  assert(is_rewrite_generation(gen));
  rewrite_gen_t ret = gen;
  switch (eviction_mode) {
  case eviction_mode_t::STOP:
    if (gen == MIN_COLD_GENERATION) {
      ret = MIN_COLD_GENERATION - 1;
    }
    break;
  case eviction_mode_t::DEFAULT:
    break;
  case eviction_mode_t::FAST:
    if (gen >= MIN_REWRITE_GENERATION && gen < MIN_COLD_GENERATION) {
      ret = MIN_COLD_GENERATION;
    }
    break;
  default:
    ceph_assert("impossible");
  }
}

// replace should_evict() and update_generation_mappings()
// for example, we can have 3 alive_ratio_threashods to control eviction modes:
// * stop_evict: 0.5
// * default_evict: 0.6
// * fast_evict: 0.7
bool check_should_evict() const {
  auto main_alive_ratio = major_cleaner->get_stat().get_alive_ratio();
  if (!has_cold_tier()) {
    eviction_mode = eviction_mode_t::STOP;
  } else {
    if (main_alive_ratio <= stop_evict) {
      eviction_mode = eviction_mode_t::STOP;
    } else if (main_alive_ratio <= default_evict) {
      if (eviction_mode > eviction_mode_t::DEFAULT) {
        eviction_mode = eviction_mode_t::DEFAULT;
      }
    } else if (main_alive_ratio <= fast_evict) {
      if (eviction_mode < eviction_mode_t::DEFAULT) {
        eviction_mode = eviction_mode_t::DEFAULT;
      }
    } else {
      assert(main_alive_ratio > fast_evict);
      eviction_mode = eviction_mode_t::FAST;
    }
  }
  return eviction_mode == eviction_mode_t::STOP;
}
```

Is it reasonable to you?

Reply:
Yeah, this implementation is more precise than mine. I'll update it.